Src/VimWpf/RelativeLineNumbers/RelativeLineNumbersMargin.cs

```diff
@@ -0,0 +1,66 @@
+﻿using System;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Text.Editor;
+
+using Vim.UI.Wpf.RelativeLineNumbers.Util;
+
+using Task = System.Threading.Tasks.Task;
+
+namespace Vim.UI.Wpf.RelativeLineNumbers
+{
+    internal sealed class RelativeLineNumbersMargin : VerticalCanvasMargin
+    {
+        private readonly IWpfTextView _textView;
+
+        private readonly ILineFormatTracker _formatTracker;
+
+        private readonly LineNumbersTracker _linesTracker;
+
+        private readonly LineNumberDrawer _lineNumberDrawer;
+
+        private readonly SafeRefreshLock _refreshLock;
+
+        private readonly LineNumbersCalculator _lineNumbersCalculator;
+
+        public override bool Enabled =>
+            _textView.Properties.GetProperty<bool>(LineNumbersMarginOptions.LineNumbersMarginOptionName);
+
+        public RelativeLineNumbersMargin(IWpfTextView textView, ILineFormatTracker formatTracker)
+            : base(LineNumbersMarginOptions.LineNumbersMarginOptionName)
+        {
+            _textView = textView
+                ?? throw new ArgumentNullException(nameof(textView));
+
+            _formatTracker = formatTracker
+                ?? throw new ArgumentNullException(nameof(formatTracker));
+
+            _lineNumbersCalculator = new LineNumbersCalculator(_textView, formatTracker);
+
+            _refreshLock = new SafeRefreshLock();
+
+            _lineNumberDrawer = new LineNumberDrawer(Canvas, _formatTracker);
+
+            _linesTracker = new LineNumbersTracker(_textView);
+
+            _linesTracker.LineNumbersChanged += async (x, y) => await RedrawLinesAsync().ConfigureAwait(true);
+            _formatTracker.VimNumbersFormatChanged += async (x, y) => await RedrawLinesAsync().ConfigureAwait(true);
+        }
+
+        private async Task RedrawLinesAsync()
+        {
+            await _refreshLock.ExecuteInLockAsync(RedrawLinesInternalAsync)
+                .ConfigureAwait(true);
+
+            async Task RedrawLinesInternalAsync()

```

Comment:
Why does this need to be async here? Feel like this is adding a bit of complication + some possible race conditions.

Reply:
I've checked for race conditions when I was writing it and it should be Ok. Initially it all was not async and was running in UI thread, but when I tried to run it in a low resources VM requests were piling up and it added to the overall delay in fast scroll scenarios. With async here we can squash all pending requests into one and make it a bit better on old PCs. It's not game-changing optimization here, but I thought that the logic of this SafeRefreshLock can be used for other operations in future to improve overall performance.